### PR TITLE
fix archive rooms becoming joined

### PR DESCRIPTION
**Pull request by nico-famedly**
When decrypting the last event for archived rooms when a room key is received, we used to send a synthetic SyncUpdate. This however put the archived room into the join section, which converted the room to a joined room. We need to respect what section the room was in when sending synthetic SyncUpdates.

fixes https://github.com/famedly/matrix-dart-sdk/issues/1916

### CODE REVIEW

**Comment by nico-famedly** [comment]
(Also includes a drive by fix for closing the db racing with running transactions)

File: lib/src/database/sqflite_box.dart [hunk 1]
```diff
@@ -66,7 +66,7 @@ class BoxCollection with ZoneTransactionMixin {
         },
       );
 
-  Future<void> close() => _db.close();
+  Future<void> close() => zoneTransaction(() => _db.close());
```
Comment (td-famedly) [inline lib/src/database/sqflite_box.dart:69]: does the indexedDb impl not need this for some reason?

Comment (nico-famedly) [reply lib/src/database/sqflite_box.dart:69]: I don't really know, it has an assertion, that it is not in a transaction...

Comment (nico-famedly) [reply lib/src/database/sqflite_box.dart:69]: Ah, those were different transactions.

Comment (td-famedly) [reply lib/src/database/sqflite_box.dart:69]: ye but the assert basically throws when we close the db while in a txn, wouldn't it be better to close the db after txn was done? (so in a zonetxn?)

File: test/room_archived_test.dart [hunk 2]
```diff
@@ -115,6 +127,66 @@ void main() {
...
+      // import the inbound session
+      await client.encryption!.keyManager.setInboundGroupSession(
+          roomid,
+          inboundSession.sessionId,
+          inboundSession.senderKey,
+          inboundSession.content);
+
+      expect(client.getArchiveRoomFromCache(roomid) != null, true);
+      expect(client.getRoomById(roomid)?.membership, Membership.leave);
+    }, skip: !encryptionSupported());
```
Comment (td-famedly) [inline test/room_archived_test.dart:188]: just tag the test as olm?

Comment (nico-famedly) [reply test/room_archived_test.dart]: Yeah, I figured that out on my own! :p

File: test/room_archived_test.dart [hunk 3]
```diff
@@ -115,6 +115,66 @@ void main() {
...
+      // import the inbound session
+      await client.encryption!.keyManager.setInboundGroupSession(
+          roomid,
+          inboundSession.sessionId,
+          inboundSession.senderKey,
+          inboundSession.content);
+
+      expect(client.getArchiveRoomFromCache(roomid) != null, true);
+      expect(client.getRoomById(roomid)?.membership, Membership.leave);
+    }, tags: 'olm');
```
Comment (td-famedly) [inline test/room_archived_test.dart:176]: you still need to init olm iirc

```dart
    setUpAll(() async {
      await olm.init();
      olm.get_library_version();
    });
    ```

Comment (nico-famedly) [reply test/room_archived_test.dart:176]: Nope, otherwise how do the tests pass? The client already initializes olm.

**Review by td-famedly — APPROVED** [review]
lg

**Pull request merged**